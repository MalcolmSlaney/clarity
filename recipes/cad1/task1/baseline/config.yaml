path:
  root: ./
  metadata_dir: ${path.root}/task1/metadata
  music_dir: ${path.root}/task1/audio/musdb18hq
  train_file: ${path.metadata_dir}/musdb18_train.json
  valid_file: ${path.metadata_dir}/musdb18_valid.demo.json
  listeners_file: ${path.metadata_dir}/listeners.demo.json
  exp_folder: ./exp # folder to store enhanced signals and final results

nalr:
  nfir: 220
  fs: 44100

compressor:
  threshold: 0.35
  attenuation: 0.1
  attack: 50
  release: 1000
  rms_buffer_size: 0.064

soft_clip: True

separator:
  device: ~

evaluate:
  set_random_seed: True
  small_test: False

# hydra config
hydra:
  run:
    dir: ${path.exp_folder}