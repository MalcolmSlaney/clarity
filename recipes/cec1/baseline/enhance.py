import json
from pathlib import Path

import hydra
import numpy as np
from omegaconf import DictConfig
from tqdm import tqdm

from clarity.enhancer.gha.gha_interface import GHAHearingAid
from clarity.evaluator.msbg.audiogram import Audiogram


@hydra.main(config_path=".", config_name="config")
def enhance(cfg: DictConfig) -> None:
    enhanced_folder = Path(cfg.path.exp_folder) / "enhanced_signals"
    enhanced_folder.mkdir(parents=True, exist_ok=True)
    with open(cfg.path.scenes_listeners_file, encoding="utf-8") as fp:
        scenes_listeners = json.load(fp)
    with open(cfg.path.listeners_file, encoding="utf-8") as fp:
        listener_audiograms = json.load(fp)

    enhancer = GHAHearingAid(**cfg["GHAHearingAid"])

    for scene in tqdm(scenes_listeners):
        for listener in scenes_listeners[scene]:
            # retrieve audiograms
            frequencies = np.array(listener_audiograms[listener]["audiogram_cfs"])
            audiogram_left_levels = np.array(
                listener_audiograms[listener]["audiogram_levels_l"]
            )
            audiogram_right_levels = np.array(
                listener_audiograms[listener]["audiogram_levels_r"]
            )
            audiogram_left = Audiogram(
                frequencies=frequencies, levels=audiogram_left_levels
            )

            audiogram_right = Audiogram(
                frequencies=frequencies,
                levels=audiogram_right_levels,
            )

            infile_names = [
                f"{cfg.path.scenes_folder}/{scene}_mixed_CH{ch}.wav"
                for ch in range(1, cfg["num_channels"] + 1)
            ]

            enhancer.process_files(
                infile_names=infile_names,
                outfile_name=f"{enhanced_folder}/{scene}_{listener}_HA-output.wav",
                audiogram_left=audiogram_left,
                audiogram_right=audiogram_right,
                listener=listener,
            )


# pylint: disable=no-value-for-parameter
if __name__ == "__main__":
    enhance()
