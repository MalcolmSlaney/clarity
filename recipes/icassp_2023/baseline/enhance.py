import json
import logging
import os
import pathlib

import hydra
import numpy as np
from omegaconf import DictConfig
from scipy.io import wavfile
from tqdm import tqdm

logger = logging.getLogger(__name__)


@hydra.main(config_path=".", config_name="config")
def enhance(cfg: DictConfig) -> None:
    enhanced_folder = pathlib.Path(cfg.path.exp_folder) / "enhanced_signals"
    os.makedirs(enhanced_folder, exist_ok=True)
    scenes_listeners = json.load(open(cfg.path.scenes_listeners_file))
    listener_audiograms = json.load(open(cfg.path.listeners_file))  # noqa: F841

    for scene in tqdm(scenes_listeners):
        fs, signal = wavfile.read(
            pathlib.Path(cfg.path.scenes_folder) / f"{scene}_mix_CH1.wav"
        )

        # Convert to 32-bit floating point scaled between -1 and 1
        signal = (signal / 32768.0).astype(np.float32)

        for listener in scenes_listeners[scene]:

            # # Audiograms can read like this, but they are not needed for the baseline
            #
            # cfs = np.array(listener_audiograms[listener]["audiogram_cfs"])
            #
            # audiogram_left = np.array(
            #    listener_audiograms[listener]["audiogram_levels_l"]
            # )
            # audiogram_right = np.array(
            #    listener_audiograms[listener]["audiogram_levels_r"]
            # )

            # Baseline just reads the signal from the front microphone pair
            # and write it out as the enhanced signal
            #

            filename = f"{scene}_{listener}_enhanced.wav"

            wavfile.write(os.path.join(enhanced_folder, filename), fs, signal)


if __name__ == "__main__":
    enhance()
