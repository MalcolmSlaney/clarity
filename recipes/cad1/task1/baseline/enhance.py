""" Run the baseline enhancement. """
from __future__ import annotations

# pylint: disable=import-error
import json
import logging
from pathlib import Path

import hydra
import numpy as np
import pandas as pd
import soxr
import torch
from omegaconf import DictConfig
from scipy.io import wavfile
from torchaudio.pipelines import HDEMUCS_HIGH_MUSDB
from torchaudio.transforms import Fade

from clarity.enhancer.compressor import Compressor
from clarity.enhancer.nalr import NALR
from clarity.utils.flac_encoder import FlacEncoder
from clarity.utils.signal_processing import denormalize_signals, normalize_signal
from recipes.cad1.task1.baseline.evaluate import make_song_listener_list

# pylint: disable=too-many-locals


logger = logging.getLogger(__name__)


def separate_sources(
    model: torch.nn.Module,
    mix: torch.Tensor | np.ndarray,
    sample_rate: int,
    segment: float = 10.0,
    overlap: float = 0.1,
    device: torch.device | str = None,
):
    """
    Apply model to a given mixture.
    Use fade, and add segments together in order to add model segment by segment.

    Args:
        model (torch.nn.Module): model to use for separation
        mix (torch.Tensor): mixture to separate, shape (batch, channels, time)
        sample_rate (int): sampling rate of the mixture
        segment (float): segment length in seconds
        overlap (float): overlap between segments, between 0 and 1
        device (torch.device, str, or None): if provided, device on which to
            execute the computation, otherwise `mix.device` is assumed.
            When `device` is different from `mix.device`, only local computations will
            be on `device`, while the entire tracks will be stored on `mix.device`.

    Returns:
        torch.Tensor: estimated sources

    Based on https://pytorch.org/audio/main/tutorials/hybrid_demucs_tutorial.html
    """
    device = mix.device if device is None else torch.device(device)
    mix = torch.as_tensor(mix, device=device)

    if mix.ndim == 1:
        # one track and mono audio
        mix = mix.unsqueeze(0)
    elif mix.ndim == 2:
        # one track and stereo audio
        mix = mix.unsqueeze(0)

    batch, channels, length = mix.shape

    chunk_len = int(sample_rate * segment * (1 + overlap))
    start = 0
    end = chunk_len
    overlap_frames = overlap * sample_rate
    fade = Fade(fade_in_len=0, fade_out_len=int(overlap_frames), fade_shape="linear")

    final = torch.zeros(batch, 4, channels, length, device=device)

    while start < length - overlap_frames:
        chunk = mix[:, :, start:end]
        with torch.no_grad():
            out = model.forward(chunk)
        out = fade(out)
        final[:, :, :, start:end] += out
        if start == 0:
            fade.fade_in_len = int(overlap_frames)
            start += int(chunk_len - overlap_frames)
        else:
            start += chunk_len
        end += chunk_len
        if end >= length:
            fade.fade_out_len = 0

    return final.cpu().detach().numpy()


def get_device(device: str) -> tuple:
    """Get the Torch device.

    Args:
        device (str): device type, e.g. "cpu", "gpu0", "gpu1", etc.

    Returns:
        torch.device: torch.device() appropiate to the hardware available.
        str: device type selected, e.g. "cpu", "cuda".
    """
    if device is None:
        if torch.cuda.is_available():
            return torch.device("cuda"), "cuda"
        return torch.device("cpu"), "cpu"

    if device.startswith("gpu"):
        device_index = int(device.replace("gpu", ""))
        if device_index > torch.cuda.device_count():
            raise ValueError(f"GPU device index {device_index} is not available.")
        return torch.device(f"cuda:{device_index}"), "cuda"

    if device == "cpu":
        return torch.device("cpu"), "cpu"

    raise ValueError(f"Unsupported device type: {device}")


def map_to_dict(sources: np.ndarray, sources_list: list[str]) -> dict:
    """Map sources to a dictionary separating audio into left and right channels.

    Args:
       sources (np.ndarray): Signal to be mapped to dictionary.
       sources_list (list): List of strings used to index dictionary.

    Returns:
        Dictionary: A dictionary of separated source audio split into channels.
    """
    audios = dict(zip(sources_list, sources))

    signal_stems = {}
    for source in sources_list:
        audio = audios[source]
        signal_stems[f"left_{source}"] = audio[0]
        signal_stems[f"right_{source}"] = audio[1]

    return signal_stems


# pylint: disable=unused-argument
def decompose_signal(
    model: torch.nn.Module,
    signal: np.ndarray,
    sample_rate: int,
    device: torch.device,
    sources_list: list[str],
    left_audiogram: np.ndarray,
    right_audiogram: np.ndarray,
    normalise: bool = True,
) -> dict[str, np.ndarray]:
    """
    Decompose signal into 8 stems.

    The left and right audiograms are ignored by the baseline system as it
    is performing personalised decomposition.
    Instead, it performs a standard music decomposition using the
    HDEMUCS model trained on the MUSDB18 dataset.

    Args:
        model_name (str): Name of the model to use. `demucs` or `openunmix`.
        model (torch.nn.Module): Torch model.
        signal (np.ndarray): Signal to be decomposed.
        sample_rate (int): Sample frequency.
        device (torch.device): Torch device to use for processing.
        sources_list (list): List of strings used to index dictionary.
        left_audiogram (np.ndarray): Left ear audiogram.
        right_audiogram (np.ndarray): Right ear audiogram.
        normalise (bool): Whether to normalise the signal.

     Returns:
         Dictionary: Indexed by sources with the associated model as values.
    """
    if normalise:
        signal, ref = normalize_signal(signal)

    sources = separate_sources(
        model, torch.from_numpy(signal), sample_rate, device=device
    )
    # only one element in the batch
    sources = sources[0]

    if normalise:
        sources = denormalize_signals(sources, ref)

    signal_stems = map_to_dict(sources, sources_list)
    return signal_stems


def apply_baseline_ha(
    enhancer: NALR,
    compressor: Compressor,
    signal: np.ndarray,
    listener_audiogram: np.ndarray,
    cfs: np.ndarray,
    apply_compressor: bool = False,
) -> np.ndarray:
    """
    Apply NAL-R prescription hearing aid to a signal.

    Args:
        enhancer: A NALR object that enhances the signal.
        compressor: A Compressor object that compresses the signal.
        signal: An ndarray representing the audio signal.
        listener_audiogram: An ndarray representing the listener's audiogram.
        cfs: An ndarray of center frequencies.
        apply_compressor: A boolean indicating whether to include the compressor.

    Returns:
        An ndarray representing the processed signal.
    """
    nalr_fir, _ = enhancer.build(listener_audiogram, cfs)
    proc_signal = enhancer.apply(nalr_fir, signal)
    if apply_compressor:
        proc_signal, _, _ = compressor.process(proc_signal)
    return proc_signal


def process_stems_for_listener(
    stems: dict,
    enhancer: NALR,
    compressor: Compressor,
    audiogram_left: np.ndarray,
    audiogram_right: np.ndarray,
    cfs: np.ndarray,
    apply_compressor: bool = False,
) -> dict:
    """Process the stems from sources.

    Args:
        stems (dict) : Dictionary of stems
        enhancer (NALR) : NAL-R prescription hearing aid
        compressor (Compressor) : Compressor
        audiogram_left (np.ndarray) : Left channel audiogram
        audiogram_right (np.ndarray) : Right channel audiogram
        cfs (np.ndarray) : Center frequencies
        apply_compressor (bool) : Whether to apply the compressor

    Returns:
        processed_sources (dict) : Dictionary of processed stems
    """

    processed_stems = {}

    for stem_str in stems:
        stem_signal = stems[stem_str]

        # Determine the audiogram to use
        audiogram = audiogram_left if stem_str.startswith("l") else audiogram_right

        # Apply NALR prescription to stem_signal
        proc_signal = apply_baseline_ha(
            enhancer, compressor, stem_signal, audiogram, cfs, apply_compressor
        )
        processed_stems[stem_str] = proc_signal
    return processed_stems


def clip_signal(signal: np.ndarray, soft_clip: bool = False) -> tuple[np.ndarray, int]:
    """Clip the signal.

    Args:
        signal (np.ndarray): Signal to be clipped and saved.
        soft_clip (bool): Whether to use soft clipping.

    Returns:
        signal (np.ndarray): Clipped signal.
        n_clipped (int): Number of samples clipped.
    """

    if soft_clip:
        signal = np.tanh(signal)
    n_clipped = np.sum(np.abs(signal) > 1.0)
    signal = np.clip(signal, -1.0, 1.0)
    return signal, int(n_clipped)


def to_16bit(signal: np.ndarray) -> np.ndarray:
    """Convert the signal to 16 bit.

    Args:
        signal (np.ndarray): Signal to be converted.

    Returns:
        signal (np.ndarray): Converted signal.
    """
    return (32767.0 * signal).astype(np.int16)


def resample(signal: np.ndarray, sample_rate: int, new_sample_rate: int) -> np.ndarray:
    """Resample the signal to the desired sample rate.

    Args:
        signal (np.ndarray): Signal to be resampled.
        sample_rate (int): Sample rate of the signal.
        new_sample_rate (int): Desired sample rate.

    Returns:
        signal (np.ndarray): Resampled signal.
    """
    if sample_rate == new_sample_rate:
        return signal
    return soxr.resample(signal, sample_rate, new_sample_rate, quality="HQ")


@hydra.main(config_path="", config_name="config")
def enhance(config: DictConfig) -> None:
    """
    Run the music enhancement.
    The system decomposes the music into vocal, drums, bass, and other stems.
    Then, the NAL-R prescription procedure is applied to each stem.
    Args:
        config (dict): Dictionary of configuration options for enhancing music.

    Returns 8 stems for each song:
        - left channel vocal, drums, bass, and other stems
        - right channel vocal, drums, bass, and other stems
    """

    if config.separator.model not in ["demucs", "openunmix"]:
        raise ValueError(f"Separator model {config.separator.model} not supported.")

    enhanced_folder = Path("enhanced_signals")
    enhanced_folder.mkdir(parents=True, exist_ok=True)

    # Training stage
    #
    # The baseline is using an off-the-shelf model trained on the MUSDB18 dataset
    # Training listeners and song are not necessary in this case.
    #
    # Training songs and audiograms can be read like this:
    #
    #  with open(config.path.listeners_train_file, "r", encoding="utf-8") as file:
    #        listener_train_audiograms = json.load(file)
    #
    #  with open(config.path.music_train_file, "r", encoding="utf-8") as file:
    #        song_data = json.load(file)
    #  songs_train = pd.DataFrame.from_dict(song_data)
    #
    # train_song_listener_pairs = make_song_listener_list(
    #     songs_train['Track Name'], listener_train_audiograms
    # )

    if config.separator.model == "demucs":
        separation_model = HDEMUCS_HIGH_MUSDB.get_model()
    else:
        separation_model = torch.hub.load("sigsep/open-unmix-pytorch", "umxhq", niter=0)
    device, _ = get_device(config.separator.device)
    separation_model.to(device)

    # Processing Validation Set
    # Load listener audiograms and songs
    with open(config.path.listeners_valid_file, encoding="utf-8") as file:
        listener_valid_audiograms = json.load(file)

    with open(config.path.music_valid_file, encoding="utf-8") as file:
        song_data = json.load(file)
    songs_valid = pd.DataFrame.from_dict(song_data)

    valid_song_listener_pairs = make_song_listener_list(
        songs_valid["Track Name"], listener_valid_audiograms
    )
    # Select a batch to process
    valid_song_listener_pairs = valid_song_listener_pairs[
        config.evaluate.batch :: config.evaluate.batch_size
    ]

    # Create hearing aid objects
    enhancer = NALR(**config.nalr)
    compressor = Compressor(**config.compressor)

    # Create flac encoder object for saving the processed stems
    flac_encoder = FlacEncoder()

    # Decompose each song into left and right vocal, drums, bass, and other stems
    # and process each stem for the listener
    prev_song_name = None
    num_song_list_pair = len(valid_song_listener_pairs)
    for idx, song_listener in enumerate(valid_song_listener_pairs, 1):
        song_name, listener_name = song_listener
        logger.info(
            f"[{idx:03d}/{num_song_list_pair:03d}] "
            f"Processing {song_name} for {listener_name}..."
        )
        # Get the listener's audiogram
        listener_info = listener_valid_audiograms[listener_name]

        # Find the music split directory
        split_directory = (
            "test"
            if songs_valid.loc[songs_valid["Track Name"] == song_name, "Split"].iloc[0]
            == "test"
            else "train"
        )

        critical_frequencies = np.array(listener_info["audiogram_cfs"])
        audiogram_left = np.array(listener_info["audiogram_levels_l"])
        audiogram_right = np.array(listener_info["audiogram_levels_r"])

        # Read the mixture signal
        # Convert to 32-bit floating point and transpose
        # from [samples, channels] to [channels, samples]
        if prev_song_name != song_name:
            # Decompose song only once
            prev_song_name = song_name

            sample_rate, mixture_signal = wavfile.read(
                Path(config.path.music_dir)
                / split_directory
                / song_name
                / "mixture.wav"
            )
            mixture_signal = (mixture_signal / 32768.0).astype(np.float32).T
            assert sample_rate == config.sample_rate

            # Model dependant params
            if config.separator.model == "demucs":
                model_sample_rate = HDEMUCS_HIGH_MUSDB.sample_rate
                sources_order = separation_model.sources
                normalise = True
            elif config.separator.model == "openunmix":
                model_sample_rate = separation_model.sample_rate
                sources_order = ["vocals", "drums", "bass", "other"]
                normalise = False
            else:
                raise ValueError(
                    f"Separator model {config.separator.model} not supported."
                )

            # Resample mixture signal to model sample rate
            if sample_rate != model_sample_rate:
                mixture_signal = resample(
                    mixture_signal, sample_rate, model_sample_rate
                )

            # Decompose mixture signal into stems
            stems = decompose_signal(
                separation_model,
                mixture_signal,
                sample_rate,
                device,
                sources_order,
                audiogram_left,
                audiogram_right,
                normalise,
            )

        # Baseline applies NALR prescription to each stem instead of using the
        # listener's audiograms in the decomposition. This stem can be skipped
        # if the listener's audiograms are used in the decomposition
        processed_stems = process_stems_for_listener(
            stems,
            enhancer,
            compressor,
            audiogram_left,
            audiogram_right,
            critical_frequencies,
            config.apply_compressor,
        )

        # save processed stems
        n_samples = processed_stems[list(processed_stems.keys())[0]].shape[0]
        out_left, out_right = np.zeros(n_samples), np.zeros(n_samples)
        for stem_str, stem_signal in processed_stems.items():
            if stem_str.startswith("l"):
                out_left += stem_signal
            else:
                out_right += stem_signal

            filename = (
                enhanced_folder
                / f"{listener_info['name']}"
                / f"{song_name}"
                / f"{listener_info['name']}_{song_name}_{stem_str}.flac"
            )
            filename.parent.mkdir(parents=True, exist_ok=True)

            # Save processed stem
            # Resample to stem sample rate.
            stem_signal = resample(
                stem_signal, config.sample_rate, config.stem_sample_rate
            )
            # Scale stem signal
            max_value = np.max(np.abs(stem_signal))
            stem_signal = stem_signal / max_value

            # Transform signal to 16-bit integer
            # and save compressed signal using flac
            flac_encoder.encode(
                to_16bit(stem_signal),
                config.stem_sample_rate,
            )

            # Save scale factor
            with open(filename.with_suffix(".txt"), "w", encoding="utf-8") as file:
                file.write(f"{max_value}")

        enhanced = np.stack([out_left, out_right], axis=1)
        filename = (
            enhanced_folder
            / f"{listener_info['name']}"
            / f"{song_name}"
            / f"{listener_info['name']}_{song_name}_remix.flac"
        )

        # Resample enhanced signal to expected output sample rate
        enhanced = resample(enhanced, config.sample_rate, config.remix_sample_rate)
        # clip and save enhanced signal
        clipped_signal, n_clipped = clip_signal(enhanced, config.soft_clip)
        if n_clipped > 0:
            logger.warning(f"Writing {filename}: {n_clipped} samples clipped")
        flac_encoder.encode(
            to_16bit(clipped_signal), config.remix_sample_rate, filename
        )


# pylint: disable = no-value-for-parameter
if __name__ == "__main__":
    enhance()
