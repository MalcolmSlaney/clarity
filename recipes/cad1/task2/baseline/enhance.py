""" Run the dummy enhancement. """
# pylint: disable=too-many-locals
# pylint: disable=import-error

import logging
from pathlib import Path
from typing import Tuple

import hydra
import numpy as np
import pyloudnorm as pyln
from omegaconf import DictConfig
from scipy.io import wavfile
from tqdm import tqdm

from recipes.cad1.task2.baseline.baseline_utils import (
    make_scene_listener_list,
    read_mp3,
)
from recipes.cad1.task2.baseline.evaluate import load_listeners_and_scenes

logger = logging.getLogger(__name__)


def enhance_song(
    waveform: np.ndarray, sample_rate: int, gain_db: float = 0.0
) -> Tuple[np.ndarray, np.ndarray]:
    """
    Enhance a single song for a listener.

    Baseline enhancement returns the original signal with a gain in dB LUFS.

    Args:
        waveform (np.ndarray): The waveform of the song.
        sample_rate (int): The sample rate of the song.
        gain_db (float): The gain to apply to the song in dB LUFS. Defaults to 0.0.

    Returns:
        out_left (np.ndarray): The enhanced left channel.
        out_right (np.ndarray): The enhanced right channel.


    Note:

    In your enhancement you may need access to listener_audiograms and config file.
    If that's the case, you can modify the function signature to include them.
    E.g.

    >> def enhance_song(
          waveform: np.ndarray,
          listener_audiograms,
          dict, cfg: DictConfig
       ) -> Tuple[np.ndarray, np.ndarray]:

    Then, left and right audiograms can be accessed as follows:

    >> left_audiogram = listener_audiograms["audiogram_levels_l"]
    >> right_audiogram = listener_audiograms["audiogram_levels_r"]

    Remember to add them to the function call in the main function.

    >> out_l, out_r = enhance_song(song_waveform, listener_audiograms, config)

    """

    if waveform.ndim == 1:
        waveform = np.array([waveform, waveform])

    meter = pyln.Meter(sample_rate)
    original_loudness = meter.integrated_loudness(waveform.T)
    waveform = pyln.normalize.loudness(
        waveform.T, original_loudness, original_loudness + gain_db
    ).T

    out_left = waveform[0, :]
    out_right = waveform[1, :]

    return out_left, out_right


@hydra.main(config_path="", config_name="config")
def enhance(config: DictConfig) -> None:
    """
    Run the music enhancement.
    The baseline system is a dummy processor that returns the input signal.

    Args:
        config (dict): Dictionary of configuration options for enhancing music.
    """
    enhanced_folder = Path("enhanced_signals") / config.evaluate.split
    enhanced_folder.mkdir(parents=True, exist_ok=True)

    # Load scenes and listeners depending on config.evaluate.split
    scenes, listener_audiograms, scenes_listeners = load_listeners_and_scenes(config)
    scene_listener_pairs = make_scene_listener_list(
        scenes_listeners, config.evaluate.small_test
    )
    scene_listener_pairs = scene_listener_pairs[
        config.evaluate.batch :: config.evaluate.batch_size
    ]

    for scene_id, listener_id in tqdm(scene_listener_pairs):
        current_scene = scenes[scene_id]
        listener = listener_audiograms[listener_id]

        song_id = current_scene["song"]
        song_path = Path(config.path.music_dir) / f"{current_scene['song_path']}"

        # Read song
        song_waveform, _ = read_mp3(song_path, config.sample_rate)
        out_l, out_r = enhance_song(song_waveform, config.sample_rate, gain_db=-5)

        enhanced = np.stack([out_l, out_r], axis=1)
        filename = f"{listener['name']}_{song_id}.wav"

        # Clip and save
        if config.soft_clip:
            enhanced = np.tanh(enhanced)
        n_clipped = np.sum(np.abs(enhanced) > 1.0)
        if n_clipped > 0:
            logger.warning(f"Writing {filename}: {n_clipped} samples clipped")
        np.clip(enhanced, -1.0, 1.0, out=enhanced)
        signal_16 = (32768.0 * enhanced).astype(np.int16)
        wavfile.write(enhanced_folder / filename, config.sample_rate, signal_16)


# pylint: disable = no-value-for-parameter
if __name__ == "__main__":
    enhance()
