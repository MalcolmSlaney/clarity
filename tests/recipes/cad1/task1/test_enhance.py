"""Tests for the enhance module"""
import numpy as np
import pytest
import torch
from torchaudio.pipelines import HDEMUCS_HIGH_MUSDB

from clarity.enhancer.compressor import Compressor
from clarity.enhancer.nalr import NALR
from recipes.cad1.task1.baseline.enhance import (
    apply_baseline_ha,
    decompose_signal,
    denormalize_signals,
    get_device,
    map_to_dict,
    normalize_signal,
    process_stems_for_listener,
    separate_sources,
)


@pytest.mark.parametrize(
    "signal,normalised_result,row_means",
    (  # Simple input
        [
            np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]),
            np.array(
                [
                    [-4.89897949, -3.67423461, -2.44948974],
                    [-1.22474487, 0.0, 1.22474487],
                    [2.44948974, 3.67423461, 4.89897949],
                ]
            ),
            np.array([4.0, 5.0, 6.0]),
        ],
        # Zero mean input
        [
            np.array([[-1, 0, 1], [-2, 0, 2], [-3, 0, 3]]),
            np.array(
                [
                    [-0.61237244, 0.0, 0.61237244],
                    [-1.22474487, 0.0, 1.22474487],
                    [-1.83711731, 0.0, 1.83711731],
                ]
            ),
            np.array([-2.0, 0.0, 2.0]),
        ],
        # Unit variance input
        [
            np.array([[0, 1, 2], [3, 4, 5], [6, 7, 8]]),
            np.array(
                [
                    [-4.89897949, -3.67423461, -2.44948974],
                    [-1.22474487, 0.0, 1.22474487],
                    [2.44948974, 3.67423461, 4.89897949],
                ]
            ),
            np.array([3.0, 4.0, 5.0]),
        ],
    ),
)
def test_normalize_signal(
    signal: np.ndarray, normalised_result: np.ndarray, row_means: np.ndarray
) -> None:
    """Test normalize_signal function"""
    result = normalize_signal(signal)
    np.testing.assert_array_almost_equal(result[0], normalised_result)
    np.testing.assert_array_almost_equal(result[1], row_means)


@pytest.mark.parametrize(
    "sources,ref,expected_result",
    (  # Test case 1: Test the function with a simple input
        [
            np.array(
                [
                    [
                        [-1.22474487, 0.0, 1.22474487],
                        [-1.22474487, 0.0, 1.22474487],
                        [-1.22474487, 0.0, 1.22474487],
                    ],
                    [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]],
                ]
            ),
            np.array([4.0, 5.0, 6.0]),
            [
                np.array([[4.0, 5.0, 6.0], [4.0, 5.0, 6.0], [4.0, 5.0, 6.0]]),
                np.array([[5.0, 5.0, 5.0], [5.0, 5.0, 5.0], [5.0, 5.0, 5.0]]),
            ],
        ],
        # Test case 2: Test the function with an input that has a non-zero mean
        [
            np.array(
                [
                    [
                        [-1.22474487, 0.0, 1.22474487],
                        [-1.22474487, 0.0, 1.22474487],
                        [-1.22474487, 0.0, 1.22474487],
                    ],
                    [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]],
                ]
            ),
            np.array([1.0, 2.0, 3.0]),
            [
                np.array([[1.0, 2.0, 3.0], [1.0, 2.0, 3.0], [1.0, 2.0, 3.0]]),
                np.array([[2.0, 2.0, 2.0], [2.0, 2.0, 2.0], [2.0, 2.0, 2.0]]),
            ],
        ],
        # Test case 3: Test the function with an input that has a non-unit variance
        [
            np.array(
                [
                    [
                        [-2.44948974, 0.0, 2.44948974],
                        [-2.44948974, 0.0, 2.44948974],
                        [-2.44948974, 0.0, 2.44948974],
                    ],
                    [[0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]],
                ]
            ),
            np.array([4.0, 5.0, 6.0]),
            [
                np.array([[3.0, 5.0, 7.0], [3.0, 5.0, 7.0], [3.0, 5.0, 7.0]]),
                np.array([[5.0, 5.0, 5.0], [5.0, 5.0, 5.0], [5.0, 5.0, 5.0]]),
            ],
        ],
    ),
)
def test_denormalize_signals(
    sources: np.ndarray, ref: np.ndarray, expected_result: np.ndarray
) -> None:
    """Test denormalize_signal function"""
    result = denormalize_signals(sources, ref)
    assert len(result) == len(expected_result)
    np.testing.assert_array_almost_equal(result[0], expected_result[0])
    np.testing.assert_array_almost_equal(result[1], expected_result[1])


def test_map_to_dict():
    sources = np.array([[1, 2], [3, 4], [5, 6]])
    sources_list = ["a", "b", "c"]
    output = map_to_dict(sources, sources_list)
    expected_output = {"l_a": 1, "r_a": 2, "l_b": 3, "r_b": 4, "l_c": 5, "r_c": 6}
    assert output == expected_output


def test_decompose_signal():
    np.random.seed(123456789)
    # Load Separation Model
    model = HDEMUCS_HIGH_MUSDB.get_model().double()
    device = torch.device("cpu")
    model.to(device)

    # Create a mock signal to decompose
    fs = 8000
    duration = 1
    signal = np.random.uniform(size=(1, 2, fs * duration))
    # Call the decompose_signal function and check that the output has the expected keys
    output = decompose_signal(model, signal, fs, device)
    expected_results = np.load(
        "../../../resources/test_enhance.test_decompose_signal.npy", allow_pickle=True
    )[()]
    for key in output.keys():
        np.testing.assert_array_almost_equal(output[key], expected_results[key])


def test_apply_baseline_ha():
    np.random.seed(987654321)
    # Create mock inputs
    signal = np.random.normal(size=44100)
    listener_audiogram = np.ones(9)
    cfs = np.array([250, 500, 1000, 2000, 4000, 6000, 8000, 9000, 10000])

    # Create mock objects for enhancer and compressor
    enhancer = NALR(nfir=220, fs=44100)
    compressor = Compressor(
        threshold=0.35, attenuation=0.1, attack=50, release=1000, rms_buffer_size=0.064
    )

    # Call the apply_nalr function and check that the output is as expected
    output = apply_baseline_ha(enhancer, compressor, signal, listener_audiogram, cfs)
    expected_results = np.load(
        "../../../resources/test_enhance.test_apply_baseline_ha.npy", allow_pickle=True
    )
    np.testing.assert_array_almost_equal(output, expected_results)


def test_process_stems_for_listener():
    np.random.seed(12357)
    # Create mock inputs
    stems = {
        "l_source1": np.random.normal(size=16000),
        "r_source1": np.random.normal(size=16000),
    }
    audiogram_left = np.ones(9)
    audiogram_right = np.ones(9)
    cfs = np.array([250, 500, 1000, 2000, 4000, 6000, 8000, 9000, 10000])

    # Create mock objects for enhancer and compressor
    enhancer = NALR(nfir=220, fs=16000)
    compressor = Compressor(
        threshold=0.35, attenuation=0.1, attack=50, release=1000, rms_buffer_size=0.064
    )

    # Call the process_stems_for_listener function and check that the output is as expected
    output = process_stems_for_listener(
        stems, enhancer, compressor, audiogram_left, audiogram_right, cfs
    )
    expected_results = np.load(
        "../../../resources/test_enhance.test_process_stems_for_listener.npy",
        allow_pickle=True,
    )[()]
    for key in output.keys():
        np.testing.assert_array_almost_equal(output[key], expected_results[key])


def test_separate_sources():
    np.random.seed(123456789)

    # Create a dummy model
    class DummyModel(torch.nn.Module):
        def __init__(self, sources):
            super().__init__()
            self.sources = sources

        def forward(self, x):
            return torch.Tensor(
                np.random.uniform(size=(x.shape[0], len(self.sources), *x.shape[1:]))
            )

    # Set up some dummy input data
    batch_size = 1
    num_channels = 1
    length = 1
    sample_rate = 16000
    sources = ["vocals", "drums", "bass", "other"]
    mix = np.random.randn(batch_size, num_channels, length * sample_rate)
    device = torch.device("cpu")

    # Create a dummy model
    model = DummyModel(sources)

    # Call separate_sources
    output = separate_sources(model, mix, sample_rate, device=device)

    expected_results = np.load(
        "../../../resources/test_enhance.test_separate_sources.npy", allow_pickle=True
    )
    # Check that the output has the correct shape
    assert output.shape == expected_results.shape
    np.testing.assert_array_almost_equal(output, expected_results)


def test_get_device():
    # Test default case (no argument passed)
    device, device_type = get_device(None)
    assert (
        device == torch.device("cuda")
        if torch.cuda.is_available()
        else torch.device("cpu")
    )
    assert device_type == "cuda" if torch.cuda.is_available() else "cpu"
