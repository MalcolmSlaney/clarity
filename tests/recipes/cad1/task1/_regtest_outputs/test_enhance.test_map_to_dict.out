output:
{'l_a': 1, 'r_a': 2, 'l_b': 3, 'r_b': 4, 'l_c': 5, 'r_c': 6}
